test/testStringFixer.js

```diff
@@ -0,0 +1,214 @@
+        test.equal(subject.ir, "\"bcdf\"");
+        test.done();
+    },
+
+    stringFixerShouldFlagAppliedFixes: function (test) {
+        test.expect(1);
+        const subject = new IntermediateRepresentation({
+            type: "string",
+            filePath: "test/file.txt",
+            ir: "abcdef",
+        });
+        const fixer = new StringFixer();
+        // no overlap
+        const fixes = [
+            // produced by rule "sentence case"
+            new StringFix(StringFixCommand.replaceAfter(0, 1, "A")),
+            // produced by rule "always shout"
+            new StringFix(StringFixCommand.insertAfter(6, "!")),
+        ];
+        fixer.applyFixes(subject, fixes);
+        test.equal(fixes.every(f => f.applied), true);
+        test.done();
+    },
+
+    stringFixerShouldSkipOverlappingFix: function (test) {
+        test.expect(2);
+        const subject = new IntermediateRepresentation({
+            type: "string",
+            filePath: "test/file.txt",
+            ir: "abcdef",
+        });
+        const fixer = new StringFixer();
+
+        // overlap
+
+        // produced by rule "always shout"
+        const alwaysShoutFix = new StringFix(StringFixCommand.insertAfter(6, "!"));
+        // produced by rule "always ask"
+        const alwaysAskFix = new StringFix(StringFixCommand.insertAfter(6, "?"));
+
+        fixer.applyFixes(subject, [alwaysShoutFix, alwaysAskFix]);
+
+        // Fixer cannot apply both fixes, because there is partial command overlap
+        // (unable to tell which is more important:
+        // should there always be exclamation mark or the question mark at the end?)
+        // so it should skip one of the fixes (the one which was further in the queue)
+        
+        test.equal(alwaysShoutFix.applied, true);
+        test.equal(alwaysAskFix.applied, false);
+
+        // @TODO note for future - this is a good example of infinite loop during naive
+        // attempt to exhaustively apply all auto-fixes by reparsing until no fixes are available:
+        // in first parsing and autofixing iteration, always-shout fix would append ! to the string
+        // and always-ask would be skipped; after reparsing, always-shout would be satisfied
+        // but always-ask would complain again, so its fix would append ?; in third iteration,
+        // always-shout would again notice missing ! at the end etc.
+        //
+        // one idea other than simply placing a hard iteration limit is:
+        // if during any iteration the rule did not produce a Result, this means that
+        // any Results produced by it in subsequent iterations would only come from errors introduced
+        // by errors in other Rules; hence, after no errors were found the Rule get disabled
+        // for the subsequent iterations
+        // or it could keep running (so that the final report would still contain a note about the issue)
+        // but become banned from producing autofixes
```

Comment:
What about this idea for the infinite loop detection algorithm? If three consecutive rounds end up with the same or greater number of Results or Fixes or if there are zero Results or Fixes, then stop, else go around again.

Reply:
Yeah that could work as well.

I actually didn't know where to note my observation so I left it as a comment here for the time being.

Should I remove it before we merge?